File: sdks/python/tests/test_codec.py
"""No-server codec round-trip tests. Run with: python -m unittest discover -s tests

These validate the byte layouts against the wire spec without needing a server,
mirroring the Rust `prism-protocol` unit tests and the Node SDK's codec tests."""

import unittest

from prismdb._codec import Reader, Writer, frame_encode
from prismdb.document import decode_document, encode_document
from prismdb.errors import ProtocolError
from prismdb.messages import MSG, decode_packet, encode_packet, sql_body
from prismdb.query import Q, encode_doc_query
from prismdb.update import U, encode_doc_update
from prismdb.value import (
    TAG,
    ObjectId,
    decode_tagged,
    encode_tagged,
    float64,
    int32,
    timestamp,
)


class CodecTest(unittest.TestCase):
    def test_writer_reader_roundtrip(self):
        w = Writer()
        w.u8(0x7F)
        w.u16(0xBEEF)
        w.u32(0xDEADBEEF)
        w.i32(-5)
        w.u64(0x1122334455667788)
        w.i64(-9000000000)
        w.f64(3.5)
        w.u128((1 << 100) + 7)
        w.str_u16("héllo")
        w.bytes_u32(b"\x00\x01\x02")
        r = Reader(w.out())
        self.assertEqual(r.u8(), 0x7F)
        self.assertEqual(r.u16(), 0xBEEF)
        self.assertEqual(r.u32(), 0xDEADBEEF)
        self.assertEqual(r.i32(), -5)
        self.assertEqual(r.u64(), 0x1122334455667788)
        self.assertEqual(r.i64(), -9000000000)
        self.assertEqual(r.f64(), 3.5)
        self.assertEqual(r.u128(), (1 << 100) + 7)
        self.assertEqual(r.str_u16(), "héllo")
        self.assertEqual(r.bytes_u32(), b"\x00\x01\x02")
        r.expect_end()

    def test_truncation_raises(self):
        r = Reader(b"\x01\x02")
        with self.assertRaises(ProtocolError):
            r.u32()

    def test_frame_encode_length_prefix(self):
        framed = frame_encode(b"abcd")
        self.assertEqual(framed[:4], (4).to_bytes(4, "little"))
        self.assertEqual(framed[4:], b"abcd")


class ValueTest(unittest.TestCase):
    def test_scalar_tagged_roundtrip(self):
        cases = [
            (None, None),
            (True, True),
            (False, False),
            (42, 42),
            (-7, -7),
            (3.25, 3.25),
            ("prism", "prism"),
            (b"\xde\xad", b"\xde\xad"),
            (int32(123), 123),
            (float64(5.0), 5.0),
            (timestamp(1_700_000_000_000_000), 1_700_000_000_000_000),
        ]
        for value, expected in cases:
            w = Writer()
            encode_tagged(w, value)
            got = decode_tagged(Reader(w.out()))
            self.assertEqual(got, expected, f"value {value!r}")

    def test_int_maps_to_int64(self):
        w = Writer()
        encode_tagged(w, 1)
        self.assertEqual(w.out()[0], TAG.INT64)

    def test_objectid_hex_roundtrip(self):
        oid = ObjectId.from_hex("507f1f77bcf86cd799439011")
        self.assertEqual(oid.to_hex(), "507f1f77bcf86cd799439011")
        self.assertEqual(str(oid), "507f1f77bcf86cd799439011")
        self.assertEqual(oid, ObjectId(oid.bytes))


class DocumentTest(unittest.TestCase):
    def test_document_roundtrip(self):
        doc = {"name": "carol", "age": 41, "active": True, "score": 9.5}
        encoded = encode_document(doc)
        # total-length prefix matches actual size
        self.assertEqual(int.from_bytes(encoded[:4], "little"), len(encoded))
        self.assertEqual(decode_document(encoded), doc)

    def test_document_rejects_binary(self):
        with self.assertRaises(ProtocolError):
            encode_document({"blob": b"\x00"})


class QueryUpdateTest(unittest.TestCase):
    def test_query_encodes_without_error(self):
        q = Q.and_(Q.eq("city", "NYC"), Q.gt("age", 30), Q.in_("tag", ["a", "b"]))
        blob = encode_doc_query(q)
        self.assertEqual(blob[0], 10)  # AND discriminant

    def test_update_encodes_count_prefix(self):
        ops = [U.set("city", "Boston"), U.inc("age", 1), U.unset("temp")]
        blob = encode_doc_update(ops)
        self.assertEqual(int.from_bytes(blob[:4], "little"), 3)


class MessageTest(unittest.TestCase):
    def test_sql_packet_header(self):
        type_code, body = sql_body("SELECT 1", [], 1)
        packet = encode_packet(7, type_code, body)
        self.assertEqual(packet[0], MSG.SQL_EXECUTE)
        self.assertEqual(int.from_bytes(packet[4:8], "little"), 7)  # request_id

    def test_decode_authack(self):
        w = Writer()
        w.u8(MSG.AUTH_ACK)
        w.raw(b"\x00\x00\x00")
        w.u32(99)  # request id echoed
        w.raw(b"\x00\x00\x00\x00")
        w.u8(0)  # status OK
        w.u64(1234)  # user_oid
        packet = decode_packet(w.out())
        self.assertEqual(packet.request_id, 99)
        self.assertEqual(packet.message.user_oid, 1234)

    def test_decode_error_trailer(self):
        w = Writer()
        w.u8(MSG.TXN_ACK)
        w.raw(b"\x00\x00\x00")
        w.u32(1)
        w.raw(b"\x00\x00\x00\x00")
        w.u8(2)  # status != 0
        w.u64(0)  # txn_id
        w.u64(0)  # commit_lsn
        w.u32(0x0201)
        w.str_u16("serialization failure")
        w.raw(b"40001")
        w.str_u16("")
        w.u32(0)
        packet = decode_packet(w.out())
        self.assertIsNotNone(packet.message.error)
        self.assertEqual(packet.message.error.code, 0x0201)
        self.assertEqual(packet.message.error.sqlstate, "40001")


if __name__ == "__main__":
    unittest.main()

File: sdks/cpp/tests/test_codec.cpp
// No-server codec round-trip tests for the C++ SDK. Header-only: needs neither
// prism.cpp nor a socket. Exits 0 if all checks pass, 1 otherwise.
//
//   g++ -std=c++17 -Iinclude tests/test_codec.cpp -o test_codec

#include "prism/prism.hpp"

#include <cstdio>

using namespace prism;

static int failures = 0;
static void check(bool cond, const char* name) {
    std::printf("%s %s\n", cond ? "ok  " : "FAIL", name);
    if (!cond) failures++;
}

static void testWriterReader() {
    detail::Writer w;
    w.u8(0x7f);
    w.u16(0xBEEF);
    w.u32(0xDEADBEEFu);
    w.i32(-5);
    w.u64(0x1122334455667788ull);
    w.i64(-9000000000ll);
    w.f64(3.5);
    w.strU16("héllo");
    w.bytesU32({0, 1, 2});

    detail::Reader r(w.out());
    check(r.u8() == 0x7f, "u8 roundtrip");
    check(r.u16() == 0xBEEF, "u16 roundtrip");
    check(r.u32() == 0xDEADBEEFu, "u32 roundtrip");
    check(r.i32() == -5, "i32 roundtrip");
    check(r.u64() == 0x1122334455667788ull, "u64 roundtrip");
    check(r.i64() == -9000000000ll, "i64 roundtrip");
    check(r.f64() == 3.5, "f64 roundtrip");
    check(r.strU16() == "héllo", "strU16 roundtrip (utf-8)");
    check(r.bytesU32() == std::vector<std::uint8_t>({0, 1, 2}), "bytesU32 roundtrip");
    r.expectEnd();
}

static void testTruncation() {
    std::vector<std::uint8_t> bytes{1, 2};
    detail::Reader r(bytes);
    bool threw = false;
    try { r.u32(); } catch (const ProtocolError&) { threw = true; }
    check(threw, "reader flags truncation");
}

static void testValueRoundtrip() {
    std::vector<Value> cases = {
        Value(),                        // null
        Value(true), Value(false),
        Value(std::int64_t(42)), Value(std::int64_t(-7)),
        Value(Int32{123}),
        Value(3.25),
        Value(std::string("prism")),
        Value(std::vector<std::uint8_t>{0xDE, 0xAD}),
        Value(Timestamp{1700000000000000ll}),
        Value(ObjectId::fromHex("507f1f77bcf86cd799439011")),
    };
    for (const auto& v : cases) {
        detail::Writer w;
        detail::encodeTagged(w, v);
        detail::Reader r(w.out());
        Value got = detail::decodeTagged(r);
        bool ok = got.tag() == v.tag();
        if (ok) {
            switch (v.tag()) {
                case Tag::Bool: ok = got.asBool() == v.asBool(); break;
                case Tag::Int32: ok = got.asInt32() == v.asInt32(); break;
                case Tag::Int64:
                case Tag::Timestamp: ok = got.asInt64() == v.asInt64(); break;
                case Tag::Double: ok = got.asDouble() == v.asDouble(); break;
                case Tag::String: ok = got.asString() == v.asString(); break;
                case Tag::Binary: ok = got.asBytes() == v.asBytes(); break;
                case Tag::ObjectId: ok = got.asObjectId() == v.asObjectId(); break;
                case Tag::Null: break;
            }
        }
        check(ok, "value tagged roundtrip");
    }
}

static void testIntDefault() {
    check(Value(1).tag() == Tag::Int64, "int literal maps to Int64");
    check(Value(Int32{1}).tag() == Tag::Int32, "Int32 wrapper maps to Int32");
}

static void testObjectId() {
    auto oid = ObjectId::fromHex("507f1f77bcf86cd799439011");
    check(oid.toHex() == "507f1f77bcf86cd799439011", "objectid hex roundtrip");
    check(oid == ObjectId(oid.bytes()), "objectid equality");
}

static void testDocument() {
    Document d{{"name", "carol"}, {"age", std::int64_t(41)}, {"active", true}, {"score", 9.5}};
    auto enc = d.encode();
    std::uint32_t total = enc[0] | (enc[1] << 8) | (enc[2] << 16) | (std::uint32_t(enc[3]) << 24);
    check(total == enc.size(), "document total-length prefix");
    Document back = Document::decode(enc.data(), enc.size());
    check(back.at("name").asString() == "carol" && back.at("age").asInt64() == 41, "document roundtrip");

    bool threw = false;
    try { Document{{"blob", std::vector<std::uint8_t>{0}}}.encode(); }
    catch (const ProtocolError&) { threw = true; }
    check(threw, "document rejects binary fields");
}

static void testQueryUpdate() {
    auto q = Query::and_(Query::eq("city", "NYC"), Query::gt("age", std::int64_t(30)),
                         Query::in("tag", {"a", "b"}));
    check(q.encode()[0] == 10, "query AND discriminant");

    Update u;
    u.set("city", "Boston").inc("age", 1).unset("temp");
    auto blob = u.encode();
    std::uint32_t count = blob[0] | (blob[1] << 8) | (blob[2] << 16) | (std::uint32_t(blob[3]) << 24);
    check(count == 3, "update count prefix");
}

static void testMessages() {
    detail::Writer body;
    body.strU32("SELECT 1");
    body.u16(0);
    body.u32(1);
    auto packet = detail::encodePacket(7, detail::msg::SqlExecute, body.out());
    check(packet[0] == detail::msg::SqlExecute, "sql packet message type");
    std::uint32_t reqId = packet[4] | (packet[5] << 8) | (packet[6] << 16) | (std::uint32_t(packet[7]) << 24);
    check(reqId == 7, "sql packet request id");

    // Build a TxnAck with an error trailer and decode it.
    detail::Writer tw;
    tw.u8(detail::msg::TxnAck);
    tw.u8(0); tw.u8(0); tw.u8(0);
    tw.u32(1);
    tw.u32(0);
    tw.u8(2);          // status != 0
    tw.u64(0);         // txn id
    tw.u64(0);         // commit lsn
    tw.u32(0x0201);
    tw.strU16("serialization failure");
    tw.raw("40001", 5);
    tw.strU16("");
    tw.u32(0);
    detail::Reader r(tw.out());
    auto h = detail::readHeader(r);
    std::uint8_t status = r.u8();
    r.u64();
    r.u64();
    auto err = detail::readTrailer(r, status);
    check(h.requestId == 1 && err && err->code == 0x0201 && err->sqlstate == "40001",
          "error trailer decode");
}

int main() {
    testWriterReader();
    testTruncation();
    testValueRoundtrip();
    testIntDefault();
    testObjectId();
    testDocument();
    testQueryUpdate();
    testMessages();
    std::printf("\n%s\n", failures == 0 ? "ALL TESTS PASSED" : "SOME TESTS FAILED");
    return failures == 0 ? 0 : 1;
}
